### PR TITLE
Add simple console exporter

### LINKED ISSUE #438: Add simple logging exporter
For testing and demonstrations it would be nice to have a simple exporter that can easily be configured to write spans to console. To make things easier, if the exporter is based on `ILogger` we can then easily write spans to any target that has an ILogger implementation, eg NLog, Log4Net, Serilog, etc.

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
Resolves #438 & #71. Super-seeds #367.

Adds a simple exporter based on `Console` that can be used to write spans. This is designed to be a learning / debugging tool and not a production exporter.

I've also added both `debug` and `experimental` to the package tags.

Based on feedback from the incomplete PR #71, I've introduced `System.Text.Json` to manage span serialisation. This does require the minimum framework the ConsoleExporter supports to be net461, but I think it's worth it as it's primarily going to be used a learning / debugging tool.

The ConsoleExporter also supports printing both compressed and indented JSON via the pretty flag.

For example
`dotnet run -p samples/Exporters/Exporters.csproj console --pretty`

### CODE REVIEW

**Comment by MikeGoldsmith** [comment]
Sure, I'll take a look.

**Comment by MikeGoldsmith** [comment]
@lmolkova I've added a generic scope before iterating the spans and another scope within the _virtual_ `WriteSpan` that scopes it to the current SpanId.

Is that along the lines of what you were thinking?

PS The `WriteSpan` method _virtual_ for easier overriding.

**Comment by lmolkova** [comment]
Sorry I didn't ask on the first review attempt: do we have a scenario for logger exporter? Why is this needed?

The reason I'm asking: there are a lot of the unknowns: do we want structured logging or string logging (i.e. everything is in a string or everything is in scope( or property on the log event) and provider can decide how to dump duration, etc). Why spanId is in scope and traceId is not? What to export and what not? Should events be another logger events in scope of the SpanContext? Basically all the questions that depend on how it is going to be used.

Generally I'm not a big fan of shipping special-case implementation where span is a string and properties have to parsed out of the string if any special-treatment is needed. I guess I can use stdout logger and forward it somewhere else if I want this feature.

Can we come up with more general case where properties of span are exposed so I can get SpanId, duration, status, etc without parsing strings. this could be achieved with [template](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-3.1#log-message-template) or scopes

**Comment by MikeGoldsmith** [comment]
I though the LoggerExporter was useful to help introduce new users to exporting, by simply connecting up to a ConsoleLogger. There is lots of benefits in being able to just print to Console or some other ILogger implementation (eg file) while testing and figuring things out.

I agree there are some big unknowns, and I just copied the LoggerExporter implementation from one of the sample projects. I expect this to be used primarily as debugging tool rather than a fully fleshed out exporter. Maybe we could improve the exporter description to indicate that. Plus the `WriteSpan` method is intentionally _virtual_ to aid with a user overriding what / how the span details are presented.

**Comment by lmolkova** [comment]
I like the intention, at the same time I believe we will end up having LoggingExporter for production scenarios and changing existing thing to satisfy production scenarios later will cost a lot along with maintaining this exporter.  
Can we keep this one in alpha/beta stage so there is no promise of non-braking behavior/API until we figure out production usage scenario?

Would stdout exporter (that I believe we have to ship for v1) solve the need for debugging exporter?

**Comment by MikeGoldsmith** [comment]
Sure, I fully understand your concerns. I could refactor the exporter to explicitly only write to Console instead of ILogger, that way we really limit the portability and have better control.

Plus I'll try to mark the project as a debugging only tool.

**Comment by lmolkova** [comment]
Thanks! BTW it seems we have an abandoned console exporter PR https://github.com/open-telemetry/opentelemetry-dotnet/pull/367 with some comments.

File: src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            this.serializerOptions.Converters.Add(new JsonStringEnumConverter());
+        }
+
+        public override Task<ExportResult> ExportAsync(IEnumerable<Span> batch, CancellationToken cancellationToken)
+        {
+            var sb = new StringBuilder();
+
+            foreach (var span in batch)
+            {
+                sb.AppendLine(JsonSerializer.Serialize(span, this.serializerOptions));
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs]: is it guaranteed to be a single line? Can we have a guaranteed single line export?

File: src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            }
+
+            System.Console.WriteLine(sb.ToString());
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs]: would serialization directly to Std.Out be more efficient? It wouldn't require allocating a whole string before writing the output.

Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs]: Also Flush after every event write may be useful.

File: src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs (hunk 1 above)
Comment (SergeyKanzhelev) [reply src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs]: single line events are handy to integrate with things like fluent.

Comment (MikeGoldsmith) [reply src/OpenTelemetry.Exporter.Console/ConsoleExporter.cs]: Yes, the compressed output remove all non-required spaces, tabs and new lines.

Example output of compressed and indented would be this:

```json
{"Context":{"TraceId":{},"SpanId":{},"TraceOptions":"Recorded","IsRemote":false,"IsValid":true,"Tracestate":[]},"Name":"child","Status":{"IsValid":true,"CanonicalCode":"Ok","Description":null,"IsOk":true},"ParentSpanId":{},"IsRecording":true,"Attributes":[{"Key":"key","Value":"value"}],"Events":[],"Links":[],"StartTimestamp":"2020-01-16T14:32:47.57907+00:00","EndTimestamp":"2020-01-16T14:32:47.580929+00:00","Kind":"Internal","LibraryResource":{"Attributes":[{"Key":"name","Value":"console-test"}]}}

{
  "Context": {
    "TraceId": {},
    "SpanId": {},
    "TraceOptions": "Recorded",
    "IsRemote": false,
    "IsValid": true,
    "Tracestate": []
  },
  "Name": "child",
  "Status": {
    "IsValid": true,
    "CanonicalCode": "Ok",
    "Description": null,
    "IsOk": true
  },
  "ParentSpanId": {},
  "IsRecording": true,
  "Attributes": [
    {
      "Key": "key",
      "Value": "value"
    }
  ],
  "Events": [],
  "Links": [],
  "StartTimestamp": "2020-01-16T14:32:54.859889+00:00",
  "EndTimestamp": "2020-01-16T14:32:54.861638+00:00",
  "Kind": "Internal",
  "LibraryResource": {
    "Attributes": [
      {
        "Key": "name",
        "Value": "console-test"
      }
    ]
  }
}
```

**Comment by MikeGoldsmith** [comment]
Updated to write each span's JSON directly to Console instead of using StringBuilder.

**Pull request merged**